### DISCUSSION TITLE
luban重构

### DISCUSSION THREAD

**Discussion by pirunxi** [focus-creative-games/luban#45] [Ideas]
我们预计将对于下半年对luban进行彻底重构，欢迎提一些改进建议及特性。

目前重构的目标为：

- 调整管线设计，分为以下几个部分：
  - 定义收集。 支持自定义，更好地适配各种风格的配置工程结构。
  - 数据加载。 基本保持目前的数据源设计。优化excel数据源格式，提供自定义微调的能力。
  - 数据校验。在当前设计基础上少量优化。
  - 代码与数据生成。基本保持目前的设计。
  - 代码与数据后处理。**新增**特性。方便对数据加密或者导入其他工作。
- 移除远程生成特性。只支持本地生成，简化代码结构。
- 管线插件化。对于管线各个部分都支持插件化，解决当前一些改动需要直接修改源码的问题，扩展新的语言或者数据格式会更容易。
- 热更新工作流优化，更好支持分包模型。
- 本地化相关优化
- 增量生成。 基本保持当前设计。
- 移除proto和db生成，将它们独立到新的项目。

**Comment by xxl-game** [comment]
校验器这里想提一个建议，能否给所有的校验器都提供单独的配置语法糖。

目前校验器的配置是直接配置在 var.type 中的，例如 

```
<var name="x1" type="int#ref=item.TbItem"/>`
```

目前只对于 ref 校验器 和 path 校验器，提供了这样的语法糖

```
<var name="x1" type="int" ref="item.TbItem"/> 等价于 
<var name="x1" type="int#ref=item.TbItem"/>

<var name="x2" type="string" path="item.TbItem"/> 等价于 
<var name="x2" type="string#path=item.TbItem"/>

```

对于其他的校验器，例如 range 校验器，size校验器 和 set 校验器，应该是没有这种语法糖的。

```
<var name="x3" type="int" range="range=[1,10]" /> 报错：定义文件包含未知属性 attr : range
```

希望 var.{{validatorName}} = xxx 这种语法糖能对所有类型的校验器都生效，这样可以方便在配置字段时把 字段类型定义 和 字段校验配置 区分开。

**Reply by pirunxi** [reply]
好的，我们考虑一下怎么做

**Comment by xxl-game** [comment]
关于校验器的另一个建议：对于 ref 引用校验，能否支持索引普通 map 表主键之外的其他子字段。

例如：

_表 A 中有字段 keyA 和 xA，
表 B 中有字段 keyB 和 xB，
keyA 和 keyB 是分别是 A，B 这两张表的主键。
需要校验：keyA 引用了 xB_

目前 ref 校验对于 map 表只能索引主键。如果需要索引其他字段，需要把该表改为 list 独立多主键表。

但我项目中的一些表，运行时仅仅需要有一个主键就够了，只在校验的时候才需要索引其他子字段。
为了 **运行前的数据校验** 而增加 **运行时的数据** 是不划算的，所以希望增加这个功能。

**Reply by pirunxi** [reply]
这个建议合理。会考虑的。

**Comment by xxl-game** [comment]
校验器建议：对于 path 校验，需要支持结尾添加 "?" 的方式忽略空白值。

ref 校验是支持忽略空白值的，例如 ```ref=item.TbItem?``` ，当单元格不填时，忽略检查。

对于 path 校验，也需要同样的功能，例如  ```path=path=Prefabs/*.prefab? ```，当单元格有填写内容时，希望检查填写的路径是否存在，当单元格不填时，忽略检查。

**Reply by xxl-game** [reply]
这个建议如果感觉合理的话，能否在近期就先添加一下。我们项目的策划最近大量使用了 luban 的校验功能，这个功能比较急需。

**Reply by pirunxi** [reply]
好的，不过可能没那么快，下半年，你可以自行改一下。

**Comment by xxl-game** [comment]
校验器建议：path 校验能否支持填写多个路径

**Reply by pirunxi** [reply]
合理。好的。
